Implement log rotation

mneuwert opened the pull request:
## Description
- Archiving logs kept for 24 hours with name ownCloud.log.<timestamp>
- Keeping up to 10 logs archived and removing older logs
- Added UI showing list of log files and allowing to delete / share individual logs

## Related Issue
#334

## Motivation and Context
- Keeping logs fairly short
- Not to clutter device storage with huge log files

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):
![Screenshot 2019-05-16 at 10 11 19](https://user-images.githubusercontent.com/228537/57837381-ffc56c80-77c2-11e9-98e9-ed840f7f4760.png)

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/owncloud/ios-app/blob/master/CONTRIBUTING.md) document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

## Test Plan:

https://github.com/owncloud/QA/blob/master/Mobile/iOS-app/Log%20Rotation.md

## Bugfixing and improvements:

- [x] (1) Option to browse hidden https://github.com/owncloud/ios-app/pull/382#issuecomment-493897390 . Moved to https://github.com/owncloud/ios-app/issues/397
- [x] (2) File deleted and not generated anymore https://github.com/owncloud/ios-app/pull/382#issuecomment-493913833 [FIXED]
- [x] (3) `Delete all` does not delete today's file https://github.com/owncloud/ios-app/pull/382#issuecomment-493917754 [FIXED]
- [x] (4) Log files have the same name https://github.com/owncloud/ios-app/pull/382#issuecomment-493922065 [FIXED]
- [x] (5) Logging disabled after disabling `Log file` https://github.com/owncloud/ios-app/pull/382#issuecomment-493927466 [FIXED]
- [x] (6) 11 log files are retained https://github.com/owncloud/ios-app/pull/382#issuecomment-493949452  [FIXED]
- [ ] (7) Remove legacy log file https://github.com/owncloud/ios-app/pull/382#issuecomment-494275301

felix-schwarz commented on ownCloud/Settings/LogFilesViewController.swift:
```diff
@@ -0,0 +1,243 @@
+		let shareButton = UIButton(type: .system)
+		shareButton.setImage(UIImage(named: "open-in"), for: .normal)
+		shareButton.addTarget(self, action: #selector(shareButtonTapped), for: .touchUpInside)
+		shareButton.frame = CGRect(origin: CGPoint(x:0.0, y:0.0), size: shareButton.imageView!.image!.size)
```
This will crash if there's no image set (i.e. when there's no longer an image named `open-in` bundled in the app). Maybe wrap the whole shareButton segment inside `if let image = UIImage(named: "open-in") {}`?

mneuwert replied:
@felix-schwarz for me it is a philosophical question: let the app crash because there is some problem and find the crash early in the QA process or may be ship not crashing app accidentally slipping through QA but containing visual glitches.

felix-schwarz commented on ownCloud/Settings/LogFilesViewController.swift:
```diff
@@ -0,0 +1,243 @@
+				self.logFileEntries.append(logEntry)
+			}
+		}
+
```
It will probably work fine the way it is, but in case APFS will not return the log files in chronological order in the future, it would be awesome to sort them by date.

Also, I feel there's an opportunity to avoid duplicating code by removing the `LogEntry` class and instead providing an equivalent from the SDK, so that `logFiles` provides an array of f.ex. `OCLogFileRecord` instances - with properties for file size, creation date and URL.

I also see the date of each log file - and especially the way to retrieve it - as an internal implementation detail. But right now there are separate implementations to get it in both the app and the SDK.

If `OCLogFileWriter.logFiles` returned an array of `OCLogFileRecord` (or any other name, really) instances with said properties, the UI would be strictly just a consumer of data provided by the SDK.

mneuwert replied:
@felix-schwarz done

felix-schwarz commented on ownCloud/Settings/LogFilesViewController.swift:
```diff
@@ -0,0 +1,243 @@
+		guard let appGroupLogsURL = OCAppIdentity.shared.appGroupLogsContainerURL else { return }
+		let logURL = appGroupLogsURL.appendingPathComponent(fileName)
+		do {
+			try FileManager.default.removeItem(at: logURL)
```
Removal, too, should be managed through the same method both in the SDK and in the app, f.ex. through a `-[OCLogWriter deleteLog:(OCLogFileRecord *)log completionHandler:(void(^)(NSError *error))completionHandler]` method. As it is, the actual implementation to delete an individual log file exists twice - once in the app and once in the SDK.

mneuwert replied:
@felix-schwarz done

michaelstingl commented:
@mneuwert add a line below to explain the retention period? (24h, 10 logs)

![2019-05-16_14_20_15](https://user-images.githubusercontent.com/214010/57853890-2dbca800-77e7-11e9-9a90-1ae4e6901fe6.png)

mneuwert commented on ownCloud/Resources/de-DE.lproj/Localizable.strings:
```diff
@@ -171,6 +171,11 @@
 /* Log settings */
+"Log Files" = "Protokolle";
+"Share" = "Teilen";
+"Delete all log files?" = "Alle Protokolle entfernnen?";
```
@felix-schwarz fixed

mneuwert commented on ownCloud/Settings/LogFilesViewController.swift:
```diff
@@ -0,0 +1,243 @@
+		OCLogger.shared.pause()
+		self.populateLogFileList()
+
+		let removeAllButtonItem = UIBarButtonItem(title: "Delete All".localized, style: .done, target: self, action: #selector(removeAllLogs))
```
@felix-schwarz fixed

mneuwert commented:
> @mneuwert add a line below to explain the retention period? (24h, 10 logs)

@michaelstingl added a note
![Screenshot 2019-05-17 at 08 58 08](https://user-images.githubusercontent.com/228537/57909036-ee8f6500-7881-11e9-9344-42f02a59cbff.png)

felix-schwarz commented on ownCloud/Settings/LogSettingsViewController.swift:
```diff
@@ -139,63 +139,13 @@ class LogSettingsViewController: StaticTableViewController {
+						self?.navigationController?.pushViewController(logFilesViewController, animated: true)
+					}, title: "Log Files".localized, accessoryType: .disclosureIndicator, identifier: "viewLogs")
+				logOutputSection?.add(row: logsRow)
+				logOutputSection?.footerTitle = "Up to 10 recent logs are kept on the device and each log can accumulate messages for the last 24 hours of the app use".localized
```
Since the File Provider also generates log output if logging is enabled, I'd like to suggest this instead:
"The last 10 log files are kept on the device - with each log file covering up to 24 hours of usage. You can export and delete log files by selecting "Browse log files"."

Regarding the "Mask private data" option, I noticed there's already some text pointing out its limitations, so I guess no change is required there.

Adding a notice below the "Enable logging" toggle might be a good idea, though, pointing out that log files
- may end up containing sensitive information
- can impact performance
- are stored on the device and not sent to ownCloud unless explicitly shared by the user themselves.
- … anything else? @michaelstingl @hosy

mneuwert replied:
@felix-schwarz I changed the first part.. And regarding adding another note to "Enable Logging" switch, I came up with following text:

_When activated, logs may impact performance and include sensitive information. However the logs are not subject to automatic submission to ownCloud servers. Sharing logs with others is sole user responsibility._

felix-schwarz commented on ownCloud/Settings/LogFilesViewController.swift:
```diff
@@ -0,0 +1,222 @@
+
+	private func populateLogFileList() {
+		guard let logFileWriter = OCLogger.shared.writer(withIdentifier: .writerFile) as? OCLogFileWriter else { return }
+		self.logRecords = logFileWriter.logRecords()
```
Maybe sort the result in reverse order so the latest log file is always on top. (I noticed you sort it from oldest to latest in the SDK as part of the log rotation routine, so the reverse sorting should likely happen in the app in order not to impact that)

mneuwert replied:
@felix-schwarz fixed

mneuwert commented on ownCloud/Settings/LogSettingsViewController.swift:
```diff
@@ -139,63 +139,13 @@ class LogSettingsViewController: StaticTableViewController {
+				let logsRow = StaticTableViewRow(subtitleRowWithAction: { [weak self] (_, _) in
+						let logFilesViewController = LogFilesViewController()
+						self?.navigationController?.pushViewController(logFilesViewController, animated: true)
+					}, title: "Log Files".localized, accessoryType: .disclosureIndicator, identifier: "viewLogs")
```
@felix-schwarz agree, changed

mneuwert commented on ownCloud/Settings/LogFilesViewController.swift:
```diff
@@ -0,0 +1,222 @@
+	override func viewWillDisappear(_ animated: Bool) {
+		super.viewWillDisappear(animated)
+		self.navigationController?.setToolbarHidden(true, animated: false)
+		NotificationCenter.default.removeObserver(self)
```
@felix-schwarz fixed

mneuwert commented on ownCloud/Settings/LogFilesViewController.swift:
```diff
@@ -0,0 +1,222 @@
+		self.toolbarItems = [flexibleSpaceButtonItem, removeAllButtonItem, flexibleSpaceButtonItem]
+		self.navigationController?.setToolbarHidden(false, animated: false)
+
+		NotificationCenter.default.addObserver(self, selector: #selector(handleLogRotationNotification), name:NSNotification.Name.OCLocalLogRotationNotificationName , object: nil)
```
@felix-schwarz fixed

felix-schwarz commented on ownCloud/Settings/LogFilesViewController.swift:
```diff
@@ -0,0 +1,235 @@
+	}
+
+	@objc private func handleLogRotationNotification() {
+		populateLogFileList()
```
Just noticed during testing: this notification can be delivered on a background thread and then lead to `tableView.reloadData()` being called on a thread different than the main thread.
```suggestion
		OnMainThread {
			self.populateLogFileList()
		}
```

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/owncloud/ios-app?pullRequest=382) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/owncloud/ios-app?pullRequest=382) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: mneuwert<br/>:x: felix-schwarz<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/owncloud/ios-app?pullRequest=382) it.</sub>

jesmrec commented:
Let's QA this

jesmrec commented:
### (1) (improvement)

1. In Settings -> Logging, disable the option "Log File"

Current: Browse log files is displayed, but no logs will be stored there anymore
Expected: Browse log option is hidden, so no file is generated. 

The special case is keep the option visible to check older logs. Not sure if this makes sense, since the logs are something that become outdated very fast

jesmrec commented:
### (2)

1. Generate some logs
2. Settings -> Logging -> Browse log files
3. Swipe in the file and remove it

Current: By generating more logs (performing actions in file list), log file is not generated again 
Expected: File is created again with new logs

iPhoneX v12.2

jesmrec commented:
### (3)

1. With logs enabled, `Delete all`

Current: Content of today's file is deleted, but a new one is created at the moment. By swiping, today's file can be deleted. 
Expected: Everything deleted in the same way if you do one by one.

This is not a bad behaviour, because today's file is deleted and generated again... so, it could be addressed separately in case you think it is out of scope.

iPhoneX v12.2

jesmrec commented:
### (4) (improvement)

Every generated file does have the same name `ownCloudApp.log`. I'd distinguish them in any way, for example, including the date, f. ex.: `ownCloudApp_20190521.log` or `ownCloudApp_21May2019.log` or similar. In case you need to send more than one, will be less confusing.

jesmrec commented:
### (5)

1. Install the app from scratch
2. Go to Settings -> Logging 
3. Disable `Log File` 
4. Kill the app, and reopen it

Current: `Enable logging` is disabled
Expected: `Enable logging` is enabled, and `Log file` disabled

iPhoneX v12.2

![May-20-2019 12-20-48](https://user-images.githubusercontent.com/14894746/58014743-cd38ae00-7af9-11e9-884f-56572c40ed1e.gif)

mneuwert commented:
> ### (3)
> 1. With logs enabled, `Delete all`
> 
> Current: Content of today's file is deleted, but a new one is created at the moment. By swiping, today's file can be deleted.
> Expected: Everything deleted in the same way if you do one by one.
> 
> This is not a bad behaviour, because today's file is deleted and generated again... so, it could be addressed separately in case you think it is out of scope.
> 
> iPhoneX v12.2

@jesmrec I think this is expected behavior, since app is constantly logging and the logger will create a new file automatically as soon as new message is logged.

jesmrec commented:
> @jesmrec I think this is expected behavior, since app is constantly logging and the logger will create a new file automatically as soon as new message is logged.

ok, i was wondering about this, but it is OK for me.

mneuwert commented:
> ### (4) (improvement)
> Every generated file does have the same name `ownCloudApp.log`. I'd distinguish them in any way, for example, including the date: `ownCloudApp_20190521.log` or `ownCloudApp_21May2019.log` or similar. In case you need to send more than one, will be less confusing.

@jesmrec The files are in fact following the following scheme ownCloud.log.<date_time>. However in the log file management UI, I just strip the timestamp. We agreed with @felix-schwarz to do so to not to clutter the UI with long file names since the creation timestamp is anyway shown below the name. When you share the file however, the timestamp will be preserved in the filename.

mneuwert commented:
> ### (1) (improvement)
> 1. In Settings -> Logging, disable the option "Log File"
> 
> Current: `Browse log files` is displayed, but no logs will be stored there anymore
> Expected: `Browse log` option is hidden, because no file is generated.
> 
> The special case is keeping the option visible to check older logs. Not sure if this makes sense, since the logs are something that become outdated very fast

@jesmrec I can do it, but it will require significant code change and will need some time to rework log settings view controller.

mneuwert commented:
> ### (2)
> 1. Generate some logs
> 2. Settings -> Logging -> Browse log files
> 3. Swipe in the file and remove it
> 
> Current: By generating more logs (performing actions in file list), log file is not generated again
> Expected: File is created again with new logs
> 
> iPhoneX v12.2

@jesmrec some how this is contradicting with (3)

jesmrec commented:
> @jesmrec The files are in fact following the following scheme ownCloud.log.<date_time>. However in the log file management UI, I just strip the timestamp. We agreed with @felix-schwarz to do so to not to clutter the UI with long file names since the creation timestamp is anyway shown below the name. When you share the file however, the timestamp will be preserved in the filename.

I was aware about the UI/UX decision not to include the timestamp but in sharing, but it was not included in file name. I have just noticed that it is included in the file name, only in case there are more than one stored log after installing. Nothing very important, but i'd include even in that case. What do you think?

jesmrec commented:
> @jesmrec some how this is contradicting with (3)

no, they are different. in (2), i do not see a log file anymore, even performing actions:

![May-20-2019 13-02-59](https://user-images.githubusercontent.com/14894746/58017106-ac735700-7aff-11e9-9c72-d282a76fc1a0.gif)

(3) is only kind of improvement, as you told, we can skip it because logs are immediately generated.

jesmrec commented:
> @jesmrec I can do it, but it will require significant code change and will need some time to rework log settings view controller.

we can add a separate issue then. This is not functionally incorrect.

jesmrec commented:
### (6)

Retain is 10 logs max

Current: 11 log files are retained:

![Screenshot 2019-05-20 at 13 34 45](https://user-images.githubusercontent.com/14894746/58018804-21489000-7b04-11e9-8f72-73fb1258816b.png)

mneuwert commented:
> ### (6)
> Retain is 10 logs max
> 
> Current: 11 log files are retained:
> 
> ![Screenshot 2019-05-20 at 13 34 45](https://user-images.githubusercontent.com/14894746/58018804-21489000-7b04-11e9-8f72-73fb1258816b.png)

@jesmrec It is 10 archived + 1 current

mneuwert commented:
> > @jesmrec The files are in fact following the following scheme ownCloud.log.<date_time>. However in the log file management UI, I just strip the timestamp. We agreed with @felix-schwarz to do so to not to clutter the UI with long file names since the creation timestamp is anyway shown below the name. When you share the file however, the timestamp will be preserved in the filename.
> 
> I was aware about the UI/UX decision not to include the timestamp but in sharing, but it was not included in file name. I have just noticed that it is included in the file name, only in case there are more than one stored log after installing. Nothing very important, but i'd include even in that case. What do you think?

@jesmrec I think I know why.. basically if you share current log, name stays unchanged. I can fix that and always add "log creation date" as suffix when sharing.

jesmrec commented:
> @jesmrec I think I know why.. basically if you share current log, name stays unchanged. I can fix that and always add "log creation date" as suffix when sharing. 

this would be an extra ball as well.

Let's summarize then:

(1) -> improvement, not to do here
(2) -> bad behaviour. it should be fixed
(3) -> expected behaviour. no fixes needed
(4) -> timestamp included with several files. Not extra work then
(5) -> bad (hidden) behaviour. it should be fixed
(6) -> 11 log files: 10 closed + 1 opened. Maybe changing the wording:  `The last 10 logs are kept...` for `The last 10 closed logs are kept...` ? (just an idea, i was confused with that)

mneuwert commented:
@jesmrec i fixed (2), added timestamp permanently to shared files (4) irrespective of if it is current or archived log. Changed wording for (6) to make it more explicit. As for (5) I was able to reproduce it only once after fresh install but reviewing code / debugging didn't bring any insight in what it could be. @felix-schwarz as you are more familiar with log level handling, do you have any ideas?

felix-schwarz commented:
Re (5) @jesmrec:

If this is the first launch, logging shouldn't be enabled as the default is that logging is off.

So I think what's happening here is this:
- the app is launched via Xcode, which probably uses the environment variable `oc:log.log-level=0`:
<img width="895" alt="Bildschirmfoto 2019-05-20 um 23 27 01" src="https://user-images.githubusercontent.com/639669/58053066-c7bd8100-7b56-11e9-99c3-a1d0d487a060.png">
- this overrides the user defaults and enabled logging at debug level directly, which helps debugging
- killing the app ends the debugging session with Xcode
- the next launch is without the environment variable set, so that now the default setting (logging = off) applies

If you either untick the checkbox in front of that environment variable in Xcode - or always launch it without Xcode, that issue should not occur.

jesmrec commented:
(2), (4), (6) -> Fixed
(5) -> by unchecking the log option in XCode, works properly. Thanks for the tip!!
(1) -> will open a new issue
(3) -> wont fix

I performed a second round with another device, everything OK

Approved

felix-schwarz commented:
### (7) Legacy log file

Didn't think about it earlier, but just noticed: if someone has been using the previous betas, the log file is in a different location and will no longer be used. Since some beta testers could have gigabytes worth of logs there, I think that the app should look for an old log file and simply delete it if one exists.

michaelstingl commented:
> I think that the app should look for an old log file and simply delete it if one exists.

👍

mneuwert commented:
> ### (7) Legacy log file
> Didn't think about it earlier, but just noticed: if someone has been using the previous betas, the log file is in a different location and will no longer be used. Since some beta testers could have gigabytes worth of logs there, I think that the app should look for an old log file and simply delete it if one exists.

@felix-schwarz Fixed with d8cf15faafd6235a63edd7dbfc68ec9f7cdaba46. Please approve if you agree with the change.

jesmrec commented:
Fix is ok from my side as well

jesmrec commented:
@michaelstingl done.
